### PR TITLE
PLAT-102259: Convert Slottable to use hooks

### PR DESCRIPTION
**Pull request by ryanjduffy**
### Checklist

* [X] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [ ] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [X] At least one test case is included for this feature or bug fix
* [X] Documentation was added or is not needed

* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)
Convert `ui/Slottable` to use hooks. Not much else to say about it. :)

### CODE REVIEW

File: packages/ui/Slottable/tests/useSlots-specs.js [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+			expect(actualErrorsLength).toBe(expectedErrorLength);
+
+			const actualError = console.error.mock.calls[0][0];
+			const expectedError = 'Warning: The slot "c" specified on div does not exist';
+
+			expect(actualError).toBe(expectedError);
+		}
+	);
+
+	test(
```
Comment (webOS101) [inline packages/ui/Slottable/tests/useSlots-specs.js:178]: Might be nice to see tests for multiple children with the same slot.

File: packages/ui/Slottable/Slottable.js [hunk 2]
```diff
@@ -103,16 +43,32 @@ const defaultConfig = {
...
+		// Slottable allows there to be other values in the destination slot and merges them.
+		// However, consumers can't avoid key warnings when merging the two lists so we should
+		// "consider this harmful" and not continue to support this with the hook and instead
+		// encourage "either/or". We should continue to support it here so this block ensures
+		// backwards compatibility.
+		slots.forEach(slot => {
+			const dist = distributed[slot];
+			const prop = rest[slot];
+			if (prop && dist) {
+				distributed[slot] = Array.isArray(dist) ? [prop, ...dist] : [prop, dist];
```
Comment (webOS101) [inline packages/ui/Slottable/Slottable.js:60]: Do we need to check if `prop` is an array as well?  Also, should we add any tests around this functionality?

Comment (webOS101) [reply packages/ui/Slottable/Slottable.js:60]: Maybe `distributed[slot] = [].concat(prop, dist)`?

Comment (webOS101) [reply packages/ui/Slottable/Slottable.js:60]: Maybe it's fine as-is.  Technically `dist` could have been a single slot that contains an array.  It could be a bit difficult to figure out.

Comment (ryanjduffy) [reply packages/ui/Slottable/Slottable.js:60]: The current code does seem to handle this scenario (surprisingly to me). I'll push an update!

Comment (ryanjduffy) [reply packages/ui/Slottable/Slottable.js]: Thanks for the tip, @webOS101! Fixed using your suggestion.

File: packages/ui/Slottable/tests/useSlots-specs.js (hunk 1 above)
Comment (ryanjduffy) [reply packages/ui/Slottable/tests/useSlots-specs.js:178]: I added a test for this as well.

**Comment by ryanjduffy** [comment]
Regarding incorporating into `kind`, generally we've tried to use `Slottable` on the outer boundary of the decorated component since it's purpose is primarily a DX improvement to source props from children and we want to operate on the "true" API everywhere else.

You're right that `useSlots` doesn't have the typical signs of a hook (state/effects). I think there is some simplicity in treating it as such anyway so that all "behaviors" are hooks but I'm not convinced it's the best approach. This _might_ be the only case where a HOC could be converted to a utility so perhaps killing the hook makes sense to avoid the exception. On the other hand, treating it as a hook now means that we could add state/effects later if we needed to without changing the API.

Thoughts?

File: packages/ui/Slottable/Slottable.js [hunk 3]
```diff
@@ -103,16 +43,32 @@ const defaultConfig = {
...
+		// Slottable allows there to be other values in the destination slot and merges them.
+		// However, consumers can't avoid key warnings when merging the two lists so we should
+		// "consider this harmful" and not continue to support this with the hook and instead
+		// encourage "either/or". We should continue to support it here so this block ensures
+		// backwards compatibility.
+		slots.forEach(slot => {
+			const dist = distributed[slot];
+			const prop = rest[slot];
+			if (prop && dist) {
+				distributed[slot] = [].concat(prop, dist);
```
Comment (Djspaceg) [inline packages/ui/Slottable/Slottable.js:60]: This is weird. I don't think I like it. I didn't even realize the old version had this behavior, and I'm probably the most frequent user of Slots. Mostly don't like it because A) I don't know that it's documented anywhere B) you can't control the order, so if it doesn't accidentally do what you want perfectly, you're SOL.

I feel like it's totally acceptable for Slottable (useSlots) to simply decide one of the two is "more valid" than the other, the slot or the prop and always prefer that one, provided it's consistent, predictable, overridable, and documented. I thin the following patterns could support those requirements with intuitive DX:

```jsx
<Component
   title="My Title"
   bodyText="potatoes are tasty"
>
   <bodyText>Only with salt on them</bodyText>
   <buttons><button>A Button!</button></buttons>
</Component>
```
The props argument can accept "fallback" values for if child-slots are not provided by the end-developer, in this example we have only props with names mapping 1:1 with existing slots.
```javascript
{children, title, bodyText} = props;
slots = ['title', 'bodyText', 'buttons'];
distributed = useSlots({slots, {children, title, bodyText}); // didn't include "buttons" prop, don't care about `buttons` prop in this case

// where distributed ==
{
   // title is fed back as a prop, so it's indistinguishable from whether it came in as a prop or a child-slot
   title: 'My Title',

   // bodyText child-slot is preferred since it seems more intentional, given that if the component author had wanted this to just be a prop, it wouldn't have been one, yet can be supported if a child-slot isn't provided, for convenience
   bodyText: 'Only with salt on them',

   // buttons comes in as a child-slot and we return it into the distributed collection. The author expects no comparison or override, since it wasn't intentionally included in the useSlots props argument.
   buttons: <button>A Button!</button>
}
```

Comment (ryanjduffy) [reply packages/ui/Slottable/Slottable.js:60]: Agreed! I'll make an update and we can review it again.

**Comment by ryanjduffy** [comment]
@webOS101 I pushed a material change to the API so I'd like to cover it with you before sending back to @Djspaceg for integration.

File: packages/ui/Slottable/useSlots.js [hunk 4]
```diff
@@ -77,36 +78,60 @@ function distribute (slots, children) {
...
 
 /**
  * Distributes `children` into the configured `slots`.
  *
+ * `useSlots` iterates over all of the `children` in `props` and distributes any children based on
+ * the followig rules:
+ *
+ * * If the child has a `slot` property matching a valid slot, or
+ * * If the component for the child has the `defaultSlot` static member matching a valid slot, or
+ * * If the child component is a string matching a valid slot.
```
Comment (webOS101) [inline packages/ui/Slottable/useSlots.js:108]: This wording is a little unclear.    'child component's type is  a string', maybe?

Comment (ryanjduffy) [reply packages/ui/Slottable/useSlots.js]: Fixed!

**Comment by ryanjduffy** [comment]
Examples of alternatives:

```js
// 1. First API but without the fallback to prop value
const {slotBefore = props.slotBefore, slotAfter = props.slotAfter, children} = useSlots({slots: ['slotBefore', 'slotAfter'], children: props.children});
```

```js
// 2. Current API with fallback to prop value
const {slotBefore, slotAfter, children} = useSlots({slots: ['slotBefore', 'slotAfter'], props});
```

The only difference between 2 and 3 is using two args vs an object arg. We had defined a default position that hooks would take a single object arg so that favors 2 but 3 is more terse. I think I prefer 2 because the object keys are a bit more clear of an API than fn args.

```js
// 3. Separate arg API with current behavior to fallback to prop value
const {slotBefore, slotAfter, children} = useSlots(['slotBefore', 'slotAfter'], props);
```

In 4, I like that it only passes the necessary prop and each prop is explicitly used triggering our lint validation. However, it's more verbose and you can't destructure both `props` and the return from useSlots without renaming things which is annoying.

```js
// 4. Explicit prop API with current behavior to fallback to prop value
const {slotBefore, slotAfter, children} = useSlots({
	slotBefore: props.slotBefore,
	slotAfter: props.slotAfter,
	children: props.children,
});
```

Using `pick()` simplifies things a bit and avoids sending all the props into `useSlots` but it requires `ramda` and is effectively the same as the current API in 2.

```js
// 5. Explicit prop API using ramda pick() with current behavior to fallback to prop value
const {slotBefore, slotAfter, children} = useSlots(pick(['slotBefore', 'slotAfter', 'children'], props));
```

I'm leaning toward the current implementation but sharing for feedback from the team first.

**Comment by webOS101** [comment]
I prefer the slightly less verbose option 3 but don't feel super-duper strongly about it.

**Comment by Djspaceg** [comment]
I think I prefer #3. Here's why.
```
// 2. Current API with fallback to prop value
const {slotBefore, slotAfter, children} = useSlots({slots: ['slotBefore', 'slotAfter'], props});
```
In the #2 version, I think it only looks simple because we happen to have named the incoming `props` argument to the wrapping function "props" and that coincides with the key that the hook is looking for. If we didn't, and did say: "rest" instead, it would look more awkward:
```
const {slotBefore, slotAfter, children} = useSlots({slots: ['slotBefore', 'slotAfter'], props: rest});
```
And we could optionally pass in different values, targeting them one at a time, which makes it look much more complicated:
```
const {slotBefore, slotAfter, children} = useSlots({slots: ['slotBefore', 'slotAfter'], props: {children: someOtherChildren, slotBefore, slotAfter: customSlotAfter}});
```
I think the _extremely mildly irksome_ thing is needing to supply that "props" key (it would be a total non-issue in most cases). The #2 approach just _seems_ more bulky. It does offer the ability to extend the API more easily over time though, which I appreciate.

What I like about #3 is it does still lend itself to that "hooks use 1 arg" mentality, essentially. The second argument, an object of props is technically optional, which I think makes good sense. It's like how the props vs props+context works. The single arg covers the primary usecase, and the bonus argument gets you _extra_ functionality. (Side benefit is we don't have to specify the word "slots" or "props", and also don't have to deal with objects within objects; it's flatter.

Ultimately I would be content with 2, 3, or even the ramda thing (once I understood it better). However, I do prefer #3.

**Comment by ryanjduffy** [comment]
The second arg in 3 is not optional. Without the props, `useSlots` can't do anything as it needs the props obj to get at `children` and the prop fallbacks.

**Comment by ryanjduffy** [comment]
I agree that the use of `props` seems easier in this simple scenario and may not pan out as such in practice.

**Comment by Djspaceg** [comment]
> The second arg in 3 is not optional. Without the props, `useSlots` can't do anything as it needs the props obj to get at `children` and the prop fallbacks.

You're right. I forgot about the passing of `children`.

**Comment by ryanjduffy** [comment]
I'm now leaning toward 4 after the discussion here. Consider the following slightly more complex scenario. We're likely to want to deconstruct in the function (or immediately after at the top of the block) and there is some readability value in using `slots` elsewhere if you're interacting with the "computed" value or spreading it onto the wrapped component if it's just a pass through.

```
function ComponentDecorator ({slotBefore, slotAfter, children, myProp, anotherProp, ...rest}) {
    const slots = useSlots({slotBefore, slotAfter, children});
    const someComputedValue = fn(myProp, anotherProp);

    return <Wrapped {...rest} {...slots} myProp={someComputedValue} />;
```

**Comment by Djspaceg** [comment]
I like this option too. I was a bit concerned about it initially since the slot names aren't explicit, but they sort of are, actually. It could lead to some interesting scenarios and code (that we, as a framework, shouldn't do internally):
```javascript
const everythingIsSlots = useSlots(props);
```
Not that we'd do that, but it technically wouldn't be forbidden.

I'm changing my answer. I like #4 the most now. :)

**Comment by webOS101** [comment]
Wow, a real world example looks so much better.  It's just what I had been hoping for but it didn't seem like it worked out.  :+1:

**Comment by ryanjduffy** [comment]
Pushed an update that uses the 4th option. I haven't checked over docs and run unit tests in sandstone yet. I'll do that tomorrow and then route back for integration.

File: packages/ui/Slottable/useSlots.js [hunk 5]
```diff
@@ -82,7 +83,7 @@ function distribute (slots, {children, ...fallback}) {
 	// will append to existing props and we want the distributed value to override the fallback
 	// value.
 	return {
-		...fallback,
+		...slots,
```
Comment (webOS101) [inline packages/ui/Slottable/useSlots.js:86]: What about deleting `slots` props whose value is `undefined`?

Comment (Djspaceg) [reply packages/ui/Slottable/useSlots.js:86]: Yeah, that's probably necessary.

File: packages/ui/Slottable/tests/useSlots-specs.js [hunk 6]
```diff
@@ -9,14 +9,15 @@ import useSlots from '../useSlots';
 describe('useSlots', () => {
 
 	test('should distribute children with a \'slot\' property', () => {
-		function Component (props) {
-			const {a, b, c} = useSlots({slots: ['a', 'b', 'c'], props});
+		// eslint-disable-next-line enact/prop-types
```
Comment (webOS101) [inline packages/ui/Slottable/tests/useSlots-specs.js:12]: Can you add a test that `children` is actually modified on output?

File: packages/ui/Slottable/useSlots.js [hunk 7]
```diff
@@ -72,11 +72,7 @@ function distribute ({children, ...slots}) {
 			}
 		});
 
-		if (remaining.length > 0) {
-			props.children = remaining;
-		} else {
-			delete props.children;
-		}
+		props.children = remaining.length === 0 ? null : remaining;
```
Comment (webOS101) [inline packages/ui/Slottable/useSlots.js:75]: I'm not sure this was strictly necessary but it's shorter than the previous so... win?

File: packages/ui/Slottable/Slottable.js [hunk 8]
```diff
@@ -62,6 +62,8 @@ const Slottable = hoc(defaultConfig, (config, Wrapped) => {
 			const prop = props[slot];
 			if (prop && dist) {
 				distributed[slot] = [].concat(prop, dist);
+			} else if (prop) {
+				distributed[slot] = prop;
```
Comment (webOS101) [inline packages/ui/Slottable/Slottable.js:66]: Why is this necessary?  Don't we spread `props` below to achieve this same effect?

Comment (ryanjduffy) [reply packages/ui/Slottable/Slottable.js:66]: that's fair!

File: packages/ui/Slottable/useSlots.js (hunk 7 above)
Comment (ryanjduffy) [reply packages/ui/Slottable/useSlots.js:77]: I've added some inline docs for this. In short, retaining `children` ensure that if the author does `<Wrapped {...props} {...distributed} />` that `props.children` is overwritten by `distributed.children` (since we already distributed the children to other props).

File: packages/ui/Slottable/Slottable.js [hunk 9]
```diff
@@ -60,7 +60,7 @@ const Slottable = hoc(defaultConfig, (config, Wrapped) => {
 		slots.forEach(slot => {
 			const dist = distributed[slot];
 			const prop = props[slot];
-			if (prop && dist) {
+			if (prop != null && dist != null) {
 				distributed[slot] = [].concat(prop, dist);
 			} else if (prop) {
```
Comment (webOS101) [inline packages/ui/Slottable/Slottable.js:65]: I think this one should be `!= null`, too.

File: packages/ui/Slottable/useSlots.js (hunk 7 above)
Comment (webOS101) [reply packages/ui/Slottable/useSlots.js:77]: That makes sense!

File: packages/ui/Slottable/Slottable.js [hunk 10]
```diff
@@ -103,16 +43,38 @@ const defaultConfig = {
...
+		const slotProps = {children: props.children};
+		slots.forEach(k => (slotProps[k] = null));
+
+		// Slottable allows there to be other values in the destination slot and merges them.
+		// However, consumers can't avoid key warnings when merging the two lists so we should
+		// "consider this harmful" and not continue to support this with the hook and instead
+		// encourage using the slot as the default with the prop as a fallback as implemented by the
+		// hook.
+		const distributed = useSlots(slotProps);
+		slots.forEach(slot => {
```
Comment (webOS101) [inline packages/ui/Slottable/Slottable.js:60]: Do we need to ensure that `children` never appears in `slots`?

File: packages/ui/Slottable/Slottable.js (hunk 9 above)
Comment (ryanjduffy) [reply packages/ui/Slottable/Slottable.js:65]: Yeah, i agree.

File: packages/ui/Slottable/Slottable.js (hunk 10 above)
Comment (ryanjduffy) [reply packages/ui/Slottable/Slottable.js:60]: We don't currently but that's a reasonable condition to add.

File: packages/ui/Slottable/tests/useSlots-specs.js (hunk 6 above)
Comment (ryanjduffy) [reply packages/ui/Slottable/tests/useSlots-specs.js:12]: Done!

File: packages/ui/Slottable/Slottable.js (hunk 10 above)
Comment (ryanjduffy) [reply packages/ui/Slottable/Slottable.js:60]: Actually, I'd like to skip this. We don't test for this currently and it isn't possible with the hook implementation so it's out of scope and really edge case-y.

Comment (webOS101) [reply packages/ui/Slottable/Slottable.js:60]: Makes sense!  Was just asking as food for thought.
